{% extends "base.html" %}
{% load static %}

{% block page_content %}
<div class="row">
    <div class="col col-md-7 col-sm-12">
      <h2>Download latest datasets</h2>
      {% comment %} <p class="notice bg-info">
        Please read <a href="doc/Manual.html">the ISDB manual</a> for guidelines on
        using HIRIS-Private and the data contained within.
      </p> {% endcomment %}

      <p>
        These downloadable datasets are broad summaries of the full data
        contained within HIRIS.
        {% comment %} Many questions are best answered by querying the database directly.

        
          <span class="visible-sm-inline visible-xs-inline">
            (See below for <a href="#access-the-database">connection details</a>.)
          </span>
        

        If you’re unfamiliar with SQL databases, you’re encouraged to
        collaborate with your local bioinformatics practitioner. {% endcomment %}
      </p>

      <p>
        Note that integration sites <strong>may be covered by multiple
        genes</strong> (such as those on opposite strands or read-through
        variants) and if so will be <strong>reported once per gene</strong>,
        resulting in multiple rows for a single integration site.  If you want
        to count integration sites, you should use the integration summary
        <em>without</em> the gene annotations.
      </p>

      <span class='data-set-limit'></span>
        <h3>Integration summary</h3>
        <p class="text-lg">
          Download latest version as:
          
            <a href="exports/integration-summary.csv" download>CSV</a>, 
          
            <a href="exports/integration-summary.xlsx" download>Excel</a>, 
           or 
            <a href="exports/integration-summary.json" download>JSON</a>
          
        </p>
        <p>
          The latest version is from <em>Thu Aug 29 07:00 2019 PDT</em>.
        </p>
        <p>This dataset contains the following fields:</p>
        <ol>
          
            <li><code>environment</code></li>
          
            <li><code>subject</code></li>
          
            <li><code>landmark</code></li>
          
            <li><code>location</code></li>
          
            <li><code>orientation_in_landmark</code></li>
          
            <li><code>multiplicity</code></li>
          
            <li><code>source_names</code></li>
          
            <li><code>pubmed_ids</code></li>
          
        </ol>
        {% comment %} <p>
          The manual contains
          <a href="doc/Manual.html#integration-summary">more information on this downloadable dataset</a>.
        </p> {% endcomment %}

        <span class='data-set-limit'></span>
        <h3>Integration summary with annotated genes</h3>
        <p class="text-lg">
          Download latest version as:
          
            <a href="exports/integration-gene-summary.csv" download>CSV</a>, 
          
            <a href="exports/integration-gene-summary.xlsx" download>Excel</a>, 
           or 
            <a href="exports/integration-gene-summary.json" download>JSON</a>
          
        </p>
        <p>
          The latest version is from <em>Thu Aug 29 07:00 2019 PDT</em>.
        </p>
        <p>This dataset contains the following fields:</p>
        <ol>
          
            <li><code>environment</code></li>
          
            <li><code>subject</code></li>
          
            <li><code>ncbi_gene_id</code></li>
          
            <li><code>gene</code></li>
          
            <li><code>gene_type</code></li>
          
            <li><code>landmark</code></li>
          
            <li><code>location</code></li>
          
            <li><code>orientation_in_landmark</code></li>
          
            <li><code>orientation_in_gene</code></li>
          
            <li><code>multiplicity</code></li>
          
            <li><code>source_names</code></li>
          
            <li><code>pubmed_ids</code></li>
          
        </ol>
        {% comment %} <p>
          The manual contains
          <a href="doc/Manual.html#integration-summary-with-annotated-genes">more information on this downloadable dataset</a>.
        </p> {% endcomment %}

        <span class='data-set-limit'></span>
        <h3>Summary by gene</h3>
        <p class="text-lg">
          Download latest version as:
          
            <a href="exports/summary-by-gene.csv" download>CSV</a>, 
          
            <a href="exports/summary-by-gene.xlsx" download>Excel</a>, 
           or 
            <a href="exports/summary-by-gene.json" download>JSON</a>
          
        </p>
        <p>
          The latest version is from <em>Thu Aug 29 07:00 2019 PDT</em>.
        </p>
        <p>This dataset contains the following fields:</p>
        <ol>
          
            <li><code>environment</code></li>
          
            <li><code>ncbi_gene_id</code></li>
          
            <li><code>gene</code></li>
          
            <li><code>gene_type</code></li>
          
            <li><code>subjects</code></li>
          
            <li><code>unique_sites</code></li>
          
            <li><code>proliferating_sites</code></li>
          
            <li><code>total_in_gene</code></li>
          
        </ol>
        {% comment %} <p>
          The manual contains
          <a href="doc/Manual.html#summary-by-gene">more information on this downloadable dataset</a>.
        </p> {% endcomment %}

    </div>
    <div class="col col-md-5 col-sm-12">
        {% comment %} <h2>Genome Browser and IGV</h2>
      <p>
        If you’re interested in a specific region of the human genome, such as
        a gene, the <a href="https://genome.ucsc.edu">UCSC Genome Browser</a>
        is a great way to start visualizing the data.  We provide two custom
        tracks which you can load into the genome browser, one for
        <a href="https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&hgt.customText=https://mullinslab.microbiol.washington.edu/hiris/private/exports/tracks/in-vivo.bed"><em>in vivo</em> sites</a>
        and another for
        <a href="https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&hgt.customText=https://mullinslab.microbiol.washington.edu/hiris/private/exports/tracks/in-vitro.bed"><em>in vitro</em> sites</a>.
      </p>
      <p>
        If you’d rather view the data in other genomics software such as
        <a href="http://software.broadinstitute.org/software/igv/home">IGV</a>,
        you can download the <a href="exports/in_vitro.bed">in vitro</a> and <a href="exports/in_vivo.bed">in vivo</a> track data as BED
        files.
      </p> {% endcomment %}

      <h3>Annotation files per gene</h3>
      <p>
        A <a href="https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md">GFF3
        file</a> is produced for each gene containing integration sites,
        letting you quickly focus on a specific gene of interest.
      </p>
      <p>
        {% comment %} Get started by <a href="doc/GFF-per-gene-exports.html">learning how to
        use these exports</a>.  Once familiar with their data, you can <a
        href="/list_gffs">browse the files</a>  {% endcomment %}
        You can <a href="/list_gffs">browse the files</a> 
        {% comment %} or <a href="exports/gff-per-gene.zip">download a zip file</a> containing them all. {% endcomment %}
      </p>
    </div>
</div>
{% endblock %}